mypy/checker.py

```diff
@@ -2964,14 +2964,16 @@ def visit_operator_assignment_stmt(self,
     def visit_assert_stmt(self, s: AssertStmt) -> None:
         self.expr_checker.accept(s.expr)
 
-        if s.msg is not None:
-            self.expr_checker.accept(s.msg)
-
         if isinstance(s.expr, TupleExpr) and len(s.expr.items) > 0:
             self.fail(message_registry.MALFORMED_ASSERT, s)
 
+        true_map, else_map = self.find_isinstance_check(s.expr)
+        if s.msg is not None:
+            with self.binder.frame_context(can_skip=False, fall_through=1):
+                self.push_type_map(else_map)
+                self.expr_checker.accept(s.msg)
```

Comment:
It looks like there is a code duplication with `ExpressionChecker.analyze_cond_branch()` why can't you re-use it here? (If there is a reason I am missing please add a comment).

Reply:
I must have missed it!  I ll check that

Reply:
It looks like there is either a misunderstanding, or you did something wrong in git. I just asked if it is possible to re-use the existing helper here, but you just removed the context manger instead, which is wrong. Namely, if the `else_map` is `None` (i.e. the condition is statically always true), then binder will mark the _current_ frame unreachable.

For example, in this test the missing return statement will be no more caught:
```
[case testAssertNew]
def f() -> int:  # Correct error here on master, no error on your branch.
    x: int
    assert isinstance(x, int), 'Hm...'
[builtins fixtures/isinstance.pyi]
```
Btw, since this was not caught by our test suite, could you please add such test? (Maybe also copy few other basic assert tests adding an assert message to them.)